### PR TITLE
Add index to wallet UTXO

### PR DESCRIPTION
**Pull request by joaopaulofonseca**
This PR implements an index of unspent transaction outputs (UTXOs).

This approach optimizes `CWallet::AvailableCoins`, by doing some of its work each time a wallet transaction changes. For instance, using `CWallet::IsMine` earlier for each transaction output, it manages to keep track only of UTXOs related to wallet addresses.

That brings large improvements on RPCs such as `listunspent` and `fundrawtransaction`. Depending on the size of your wallet and number of UTXOs, performance gains may be up to 98%.

This way, instead of going through all the wallet transactions, the search for UTXOs is made only on this updated list of outputs.

Depends on #7822 to ensure functionality behaviour remains the same.

### CODE REVIEW

**Comment by laanwj** [comment]
>  Depending on the size of your wallet and number of UTXOs, performance gains may be up to 98%.

Yes, this makes sense.

How does it affect memory usage?

**Comment by sipa** [comment]
Concept ACK, it's silly that we didn't have this.

Memory usage: I expect it to consume around 80 bytes per unspent output.

**Comment by promag** [comment]
ACK

Rename first commit to something like `Add index to wallet UTXO`

**Comment by jpdffonseca** [comment]
The overall memory usage will depend on the wallet size, more particularly on the UTXO count. Given the info being stored per UTXO (an outpoint and a pair with `isminetype` and pointer to Tx), the extra memory increment won't be that much.

**Comment by jpdffonseca** [comment]
@promag that makes sense. This is not a cache, but more like an index..

**Comment by pedrobranco** [comment]
Tested ACK.

In my case the improvements of listunspent are from 12s. to **0.1s** .

**Comment by jonasschnelli** [comment]
Nice work!
Concept ACK.

**Comment by laanwj** [comment]
> Given the info being stored per UTXO (an outpoint and a pair with isminetype and pointer to Tx), the extra memory increment won't be that much.

So this only uses (~80 bytes) memory per actually unspent output that is in the wallet? That's great. I then don't expect the extra memory use for that to be significant. The bulk is wasted on keeping historical transactions anyway.

Concept ACK.

**Comment by gmaxwell** [comment]
Concept ACK. Thanks for working on this.

**Comment by promag** [comment]
Solves #6573.

**Comment by instagibbs** [comment]
lightly tested ACK

**Comment by promag** [comment]
This change greatly improves the performance of AvailableCoins and indirectly listunspent and others. Is it reasonable to expect this in next 0.12 update?

**Comment by sipa** [comment]
@promag Only bug fixes are backported to old releases.

**Comment by laanwj** [comment]
I think this needs at least one more code review ACK before merge (to master, not 0.12)

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -498,9 +502,12 @@ bool CWallet::IsSpent(const uint256& hash, unsigned int n) const
 
 void CWallet::AddToSpends(const COutPoint& outpoint, const uint256& wtxid)
 {
+    // since this output is being marked as a spend output, we remove it from the UTXO index
```
Comment (sipa) [inline src/wallet/wallet.cpp]: Typo: spent output

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -517,6 +524,49 @@ void CWallet::AddToSpends(const uint256& wtxid)
...
+            if (!filterIsMine || ismine) {
+                mapTxUnspents[outpoint] = std::make_pair(wtx, ismine);
+            }
+        }
+    }
+}
+
+void CWallet::RemoveFromUnspents(const COutPoint &outpoint)
+{
+    TxUnspents::iterator uit = mapTxUnspents.find(outpoint);
```
Comment (sipa) [inline src/wallet/wallet.cpp]: Simpler to write as mapTxUnspents.erase(outpoint)

**Comment by sipa** [comment]
How does this deal with:
* reorganizations (coins marked as spent may go back to being unspent)?
* conflicts (there may be multiple unconfirmed transactions at the tip, and which set of them is accepted can change easily)?
* abandoned transactions?

**Comment by fanquake** [comment]
This needs a rebase.

**Comment by sipa** [comment]
@jpdffonseca Feel like answering my questions above?

**Comment by jpdffonseca** [comment]
@sipa I'm working to unify `mapTxSpends` and `mapTxUnspents`.

**Comment by paveljanik** [comment]
@jpdffonseca Are you still working on this? We need performance improvements especially for large wallets.

Needs rebase.

**Comment by jpdffonseca** [comment]
@paveljanik lately I have not had much time to work on it. After testing more deeply, I realized that this solution does not always handles well some of the cases mentioned above by @sipa.

So I've been trying another improvements that optimize not the `unspents` list itself but other values necessary to calculate the `unspents`, such as the call `CWallet::IsMine()`. 
With that I obtained satisfactory results, around 50% for large wallets. I'll finish that and submit PR.

**Comment by fanquake** [comment]
Closing due to inactivity. @joaopaulofonseca if you do end up finishing work on your other improvements PRs are welcome.

**Comment by joaopaulofonseca** [comment]
Deprecated partially by #9939. Caching accurately these values gets a little tricky because of all the scenarios that could make the cache invalidated (conflicts, double-spends, reorganizations, abandoned transactions, etc, ...).

**Pull request closed without merging**